File: sync-i18n.js
const fs = require('fs');

// 1. Quelldateien einlesen (die festen Konstanten und die Schichten)
const shiftsData = JSON.parse(fs.readFileSync('./shifts.json', 'utf8'));
const baseDe = JSON.parse(fs.readFileSync('./i18n/de.base.json', 'utf8'));
const baseEn = JSON.parse(fs.readFileSync('./i18n/en.base.json', 'utf8'));

// 2. Neue Ziel-Objekte erstellen (Kopie der Konstanten als Startpunkt)
const finalDe = { ...baseDe };
const finalEn = { ...baseEn };

if (shiftsData.eventI18nKey && shiftsData.eventTranslations) {
    // Event Name
    if (shiftsData.eventTranslations.name) {
        if (shiftsData.eventTranslations.name.de) finalDe[`${shiftsData.eventI18nKey}_name`] = shiftsData.eventTranslations.name.de;
        if (shiftsData.eventTranslations.name.en) finalEn[`${shiftsData.eventI18nKey}_name`] = shiftsData.eventTranslations.name.en;
    }
    // Event Beschreibung
    if (shiftsData.eventTranslations.desc) {
        if (shiftsData.eventTranslations.desc.de) finalDe[`${shiftsData.eventI18nKey}_desc`] = shiftsData.eventTranslations.desc.de;
        if (shiftsData.eventTranslations.desc.en) finalEn[`${shiftsData.eventI18nKey}_desc`] = shiftsData.eventTranslations.desc.en;
    }
}

// 3. Variable Daten aus den Schichten hinzufügen
shiftsData.eventTasks.forEach(task => {
    if (task.i18nKey && task.translations) {
        
        // Task Name
        if (task.translations.name) {
            if (task.translations.name.de) finalDe[`${task.i18nKey}_name`] = task.translations.name.de;
            if (task.translations.name.en) finalEn[`${task.i18nKey}_name`] = task.translations.name.en;
        }

        // Task Beschreibung
        if (task.translations.desc) {
            if (task.translations.desc.de) finalDe[`${task.i18nKey}_desc`] = task.translations.desc.de;
            if (task.translations.desc.en) finalEn[`${task.i18nKey}_desc`] = task.translations.desc.en;
        }
    }

    // Schichten innerhalb des aktuellen Tasks durchlaufen
    if (task.taskShifts) {
        task.taskShifts.forEach(shift => {
            if (shift.i18nKey && shift.translations) {
                
                // Schicht Name (z.B. "10:00 - Ende (mit Auto)")
                if (shift.translations.name) {
                    if (shift.translations.name.de) finalDe[`${shift.i18nKey}_name`] = shift.translations.name.de;
                    if (shift.translations.name.en) finalEn[`${shift.i18nKey}_name`] = shift.translations.name.en;
                }

                // Schicht Beschreibung (für den optionalen blauen Chip)
                if (shift.translations.desc) {
                    if (shift.translations.desc.de) finalDe[`${shift.i18nKey}_desc`] = shift.translations.desc.de;
                    if (shift.translations.desc.en) finalEn[`${shift.i18nKey}_desc`] = shift.translations.desc.en;
                }
            }
        });
    }
});

// 4. Die finalen, zusammengefassten Dateien generieren
fs.writeFileSync('./i18n/de.json', JSON.stringify(finalDe, null, 4));
fs.writeFileSync('./i18n/en.json', JSON.stringify(finalEn, null, 4));

console.log('🚀 Erfolgreich: de.json und en.json wurden aus den Basis-Dateien und shifts.json neu generiert!');

File: sync-i18n.py
import json

def read_json(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        return json.load(f)

# read source data (constant base.json and variable shifts.json)
shiftsData = read_json('./shifts.json')
baseDe = read_json('./i18n/de.base.json')
baseEn = read_json('./i18n/en.base.json')

# copy base.json
finalDe = baseDe.copy()
finalEn = baseEn.copy()

event_i18n_key = shiftsData.get("eventI18nKey")
event_translations = shiftsData.get("eventTranslations")
if event_i18n_key and event_translations:
    # event name
    trans_name = event_translations.get('name')
    if trans_name:
        if trans_name.get("de"): finalDe[f"{event_i18n_key}_name"] = trans_name['de']
        if trans_name.get("en"): finalEn[f"{event_i18n_key}_name"] = trans_name['en']
    
    # event descriptions
    trans_desc = event_translations.get('desc')
    if trans_desc:
        if trans_desc.get("de"): finalDe[f"{event_i18n_key}_desc"] = trans_desc['de']
        if trans_desc.get("en"): finalEn[f"{event_i18n_key}_desc"] = trans_desc['en']

# add variable data from shifts.json
eventTasks = shiftsData.get('eventTasks', [])
for task in eventTasks:
    # add tasks
    task_i18n_key = task.get("i18nKey")
    task_translations = task.get("translations")
    if task_i18n_key and task_translations:
        # task name
        trans_name = task_translations.get("name")
        if trans_name:
            if trans_name.get("de"): finalDe[f"{task_i18n_key}_name"] = trans_name["de"]
            if trans_name.get("en"): finalEn[f"{task_i18n_key}_name"] = trans_name["en"]

        # task description
        trans_desc = task_translations.get("desc")
        if trans_desc:
            if trans_desc.get("de"): finalDe[f"{task_i18n_key}_desc"] = trans_desc["de"]
            if trans_desc.get("en"): finalEn[f"{task_i18n_key}_desc"] = trans_desc["en"]


    taskShifts = task.get("taskShifts", [])
    for shift in taskShifts:
        # add shifts
        shift_i18n_key = shift.get("i18nKey")
        shift_translations = shift.get("translations")
        if shift_i18n_key and shift_translations:
                # shift name
                trans_name = shift_translations.get("name")
                if trans_name:
                    if trans_name.get("de"): finalDe[f"{shift_i18n_key}_name"] = trans_name["de"]
                    if trans_name.get("en"): finalEn[f"{shift_i18n_key}_name"] = trans_name["en"]

                # shift description
                trans_desc = shift_translations.get("desc")
                if trans_desc:
                    if trans_desc.get("de"): finalDe[f"{shift_i18n_key}_desc"] = trans_desc["de"]
                    if trans_desc.get("en"): finalEn[f"{shift_i18n_key}_desc"] = trans_desc["en"]

with open('./i18n/de.json', 'w', encoding='utf-8') as f:
    json.dump(finalDe, f, indent=4, ensure_ascii=False)
with open('./i18n/en.json', 'w', encoding='utf-8') as f:
    json.dump(finalEn, f, indent=4, ensure_ascii=False)

print('success: translation files updated')